### PR TITLE
Simplify HTTP Interceptors BBE

### PR DESCRIPTION
**Pull request by TharmiganK**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: examples/http-response-interceptors/http_response_interceptors.bal [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import ballerina/http;
+
+// Header name to be set to the response in the response interceptor.
+final string interceptor_header = "responseHeader";
```
Comment (shafreenAnfar) [inline examples/http-response-interceptors/http_response_interceptors.bal:4]: Do we still need them ?

File: examples/http-response-interceptors/http_response_interceptors.bal [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+// Creates a new `ResponseInterceptor`.
+ResponseInterceptor responseInterceptor = new;
+
+// Engage interceptors at the listener level. Response interceptor services will be executed from
+// tail to head.
+listener http:Listener interceptorListener = new http:Listener(9090,
+    // This interceptor pipeline will be executed for all of the services attached to this listener.
+    interceptors = [responseInterceptor]
```
Comment (shafreenAnfar) [inline examples/http-response-interceptors/http_response_interceptors.bal:36]: we can just do `[new ResponseInterceptor()]`

File: examples/http-response-interceptors/http_response_interceptors.bal [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+);
+
+service /user on interceptorListener {
+
+    resource function get greeting(http:Request req) returns http:Ok|error {
+        return {
+            mediaType: "application/org+json",
```
Comment (shafreenAnfar) [inline examples/http-response-interceptors/http_response_interceptors.bal:43]: returning text/plain simplifies the code right

Comment (TharmiganK) [reply examples/http-response-interceptors/http_response_interceptors.bal]: Fixed in [4ab29b4](https://github.com/ballerina-platform/ballerina-distribution/pull/3032/commits/4ab29b4d2494a5462fdfe7311043e16b52c30282)

File: examples/http-response-interceptors/http_response_interceptors.bal (hunk 2 above)
Comment (TharmiganK) [reply examples/http-response-interceptors/http_response_interceptors.bal]: Fixed in [4ab29b4](https://github.com/ballerina-platform/ballerina-distribution/pull/3032/commits/4ab29b4d2494a5462fdfe7311043e16b52c30282)

File: examples/http-response-interceptors/http_response_interceptors.bal (hunk 1 above)
Comment (TharmiganK) [reply examples/http-response-interceptors/http_response_interceptors.bal:4]: This variable is just there to set a header to the response inside a response interceptor. That is the only indication of  the response interceptor execution.

File: examples/http-request-interceptors/http_request_interceptors.bal [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        req.setHeader(interceptor_header1, interceptor_header_value1);
+        // Returns the next interceptor or the target service in the pipeline. 
+        // An error is returned when the call fails.
+        return ctx.next();
+    }
+}
+
+// An `Interceptor` service class with a specific path. This interceptor can only be 
+// engaged at the service level.
+service class RequestInterceptor2 {
```
Comment (chamil321) [inline examples/http-request-interceptors/http_request_interceptors.bal:31]: IMO one interceptor is enough. We can explain this filtering behaviour by improving the comments of `RequestInterceptor1`
Say that `By specifying the accessor and path in the interceptor resource, requests can be filtered to intercept only the matched ones` or something similar to that?

Comment (TharmiganK) [reply examples/http-request-interceptors/http_request_interceptors.bal]: +1

File: examples/http-request-interceptors/http_request_interceptors.bal [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+// A `Requestinterceptorservice` class implementation. It intercepts the request
+// and adds a header before it is dispatched to the target service. A `RequestInterceptorService`
+// class can have only one resource function. 
+service class RequestInterceptor {
+    *http:RequestInterceptor;
+
+    // A default resource function, which will be executed for all the requests. 
+    // A `RequestContext` is used to share data between the interceptors.
+    // An accessor and a path can also be specified. In that case, the interceptor will be
+    // executed only for the requests which match the accessor and path.
```
Comment (praneesha) [inline examples/http-request-interceptors/http_request_interceptors.bal:18]:
```suggestion
    // executed only for the requests, which match the accessor and path.
```

File: examples/http-request-interceptors/http_request_interceptors.description [hunk 6]
```diff
@@ -0,0 +1,6 @@
+// Interceptors are used to execute some common logic such as logging, header manipulation,
+// state publishing, etc. for all the inbound requests and outbound responses. A 
+// `RequestInterceptor` can be used to intercept the request and execute some custom
+// logic. `RequestInterceptors` have a resource method, which will be executed
+// before dispatching the request to the actual resource in the target service.
+// For more information, see the [HTTP module](https://docs.central.ballerina.io/ballerina/http/latest/).
```
Comment (praneesha) [inline examples/http-request-interceptors/http_request_interceptors.description:6]:
```suggestion
// For more information, see the [`http` module](https://docs.central.ballerina.io/ballerina/http/latest/).
```

File: examples/http-request-interceptors/http_request_interceptors.md [hunk 7]
```diff
@@ -0,0 +1,19 @@
+# Request Interceptors
```
Comment (praneesha) [inline examples/http-request-interceptors/http_request_interceptors.md:1]:
```suggestion
# Request interceptors
```

File: examples/http-request-interceptors/http_request_interceptors.md [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+Interceptors are used to execute some common logic such as logging, header manipulation,
+state publishing, etc. for all the inbound requests and outbound responses. A 
+`RequestInterceptor` can be used to intercept the request and execute some custom
+logic. `RequestInterceptors` have a resource method, which will be executed
+before dispatching the request to the actual resource in the target service.
+
+For more information, see the [HTTP module](https://docs.central.ballerina.io/ballerina/http/latest/).
```
Comment (praneesha) [inline examples/http-request-interceptors/http_request_interceptors.md:9]:
```suggestion
For more information, see the [`http` module](https://docs.central.ballerina.io/ballerina/http/latest/).
```

File: examples/http-response-interceptors/http_response_interceptors.bal [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+// Header name to be set to the response in the response interceptor.
+final string interceptor_header = "responseHeader";
+
+// Header value to be set to the response in the response interceptor.
+final string interceptor_header_value = "ResponseInterceptor";
+
+// A `ResponseInterceptor` service class implementation. It intercepts the response 
+// and adds a header before it is dispatched to the client. A `ResponseInterceptor`
+// service class can have only one remote function : `interceptResponse`.
```
Comment (praneesha) [inline examples/http-response-interceptors/http_response_interceptors.bal:11]:
```suggestion
// service class can have only one remote function: `interceptResponse`.
```

File: examples/http-response-interceptors/http_response_interceptors.bal [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+service class ResponseInterceptor {
+    *http:ResponseInterceptor;
+
+    // The `interceptResponse` remote function, which will be executed for all the
```
Comment (praneesha) [inline examples/http-response-interceptors/http_response_interceptors.bal:15]:
```suggestion
    // The `interceptResponse` remote function will be executed for all the
```

File: examples/http-response-interceptors/http_response_interceptors.md [hunk 11]
```diff
@@ -0,0 +1,18 @@
+# Response Interceptors
+
+In addition to `RequestInterceptors`, a `ResponseInterceptor` can be used to intercept the response.
+`ResponseInterceptors` have a remote method, which will be executed before dispatching the response 
+to the client. A collection of these request and response interceptors can be configured as a pipeline
+at the listener level or service level. 
+
+For more information, see the [HTTP module](https://docs.central.ballerina.io/ballerina/http/latest/).
```
Comment (praneesha) [inline examples/http-response-interceptors/http_response_interceptors.md:8]:
```suggestion
For more information, see the [`http` module](https://docs.central.ballerina.io/ballerina/http/latest/).
```

**Review by praneesha — APPROVED** [review]

**Pull request merged**